### PR TITLE
Reset auto increment sequence

### LINKED ISSUE #6321: Reset sequence value after file-based import
adding a new item on a collection, get the following error 

There were validation errors for the following hidden fields:
id:  Value has to be unique 

Version: Directus v9.0.0-rc.76

this seemed to have started after I updated some items with the new import from file method but I am not sure.

### PR DESCRIPTION
**Pull request by jaads**
## Fixes https://github.com/directus/directus/issues/6321

With this the database sequence for the auto_increment value is being reset, when an item with a manual primary key gets inserted in the database. Providing the primary keys manually, although the PK column is configured to automatically increment, caused an error in PostgreSQL. The error appeared when the automatically calculated key by the database collided with the manually provided key. Since cloud users cannot reset the sequence themself, this needs to be done automatically. 

Other databases handle this automatically, so the auto generated PK is always the MAX value of the column and hence no collision appeared. This includes mysql, mariadb and sqlite. Cockroachdb, mssql and oracle db on the other hand don't do this automatically. However to target those databases more work is required - time we currently don't have. Luckily in mssql  no manual key can provided - it errors unless this option is changed.  

The sequence reset gets triggered 
- after a single item is created with a manual primary key or
- or after multiple items which all have a PK given, were inserted

## Example for a single, flat item
let's assume we want to add two items. One with a manual primary key ( `id`) and the other without an manual pk.
it doesn't matter, if those item get inserted with a single request or two separate requests. 
```json
{
    "something": "foo",
    "id": 10
}
```
and one without a PK
```json
{
    "something": "foo"
}
```

let's assume the current state
```json
{
    "data": [
        {
            "id": 1,
            "something": "foo"
        },
        {
            "id": 2,
            "something": "foo"
        }
    ]
}
```

### Before
After the first item the state would look like this
```json
{
    "data": [
        {
            "id": 1,
            "something": "foo"
        },
        {
            "id": 2,
            "something": "foo"
        },
        {
            "id": 10,
            "something": "foo"
        }
    ]
}
```
Note the gap between the last auto incremented value (2) and the manually provided key (10).
Without the change, the next value of the auto increment sequence would be **3**: 
```json
{
    "data": [
        {
            "id": 1,
            "something": "foo"
        },
        {
            "id": 2,
            "something": "foo"
        },
    {
            "id": 3,
            "something": "foo"
        }
        {
            "id": 10,
            "something": "foo"
        }
    ]
}
```
So after a few other insertion with auto incremented value, the value would collide with the manual provided key `10`.

### After
Now when a manual key is provided, the auto increment sequence will get reset.
Now with the fix, when inserting the first item, the sequence will get reset for postgres to the max value of the pk column, which would be 10. Inserting another item without a PK would then have a PK of 11 and no collision will happen.
```json
{
    "data": [
        {
            "id": 1,
            "something": "foo"
        },
        {
            "id": 2,
            "something": "foo"
        },
        {
            "id": 10,
            "something": "foo"
        },
  {
            "id": 11,
            "something": "foo"
        }
    ]
}
```

## TODO

- [x] Blackbox tests will be added soon (@licitdev will have a look)

### CODE REVIEW

**Review by joselcvarela — DISMISSED** [review]
Logic needs to be done whenever an item is created, and this PR only taking into account:
- When item/items is/are created through POST /items/<collection>

But we still need to figure:
- When item/items is/are created through relationship
- When item/items is/are created through import utility
- When item/items is/are created directly into database

Also, what should happen when we have a collection with millions of records and we delete all those records? We probably want to reset sequence when we also delete right?

File: api/src/database/helpers/sequence/dialects/postgres.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import { AutoSequenceHelper } from '../types.js';
+
+export class AutoIncrementHelperPostgres extends AutoSequenceHelper {
+	/**
+	 * Resets the auto increment sequence for a table based on the max value of the PK column.
+	 * We're assuming that the default sequence name is being used,
+	 * which is the `${tableName}_${columnName}_seq`.
+	 */
+	override resetAutoIncrementSequence(_table: string, _column: string): Knex.Raw | null {
+		return this.knex.raw(`SELECT SETVAL('${_table}_${_column}_seq', (SELECT MAX(${_column}) FROM ${_table}));`);
```
Comment (licitdev) [inline api/src/database/helpers/sequence/dialects/postgres.ts:11]: We can fetch the sequence name to eliminate the assumption. 
Here's an example:
```sql
WITH info AS (
    SELECT
        pg_get_serial_sequence('table_name', 'pk_field') AS seq_name,
        MAX(pk_field) as max_val
    FROM
        table_name
)
SELECT
    SETVAL(seq_name, max_val)
FROM
    info;
```

File: api/src/utils/reset-sequence.ts
Comment (licitdev) [inline api/src/utils/reset-sequence.ts:1]: I think we should rename the file or function to match for consistency. 🙊

File: api/src/database/helpers/sequence/dialects/postgres.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import type { Knex } from 'knex';
+import { AutoSequenceHelper } from '../types.js';
+
+export class AutoIncrementHelperPostgres extends AutoSequenceHelper {
+	/**
+	 * Resets the auto increment sequence for a table based on the max value of the PK column.
+	 * We're assuming that the default sequence name is being used,
+	 * which is the `${tableName}_${columnName}_seq`.
+	 */
+	override resetAutoIncrementSequence(_table: string, _column: string): Knex.Raw | null {
```
Comment (licitdev) [inline api/src/database/helpers/sequence/dialects/postgres.ts:10]: The underscore in the `_table` and `_column` parameters should be removed as they are not unused.

**Review by licitdev — CHANGES_REQUESTED** [review]
Adding some comments from an initial review. Nice use of helpers for vendor specific logic. 👍

Great points from @joselcvarela. If this logic is implemented within `ItemsService`, most cases from nested relational items to the import utility should be covered. 🤔 

I think that we should not reset sequences automatically on deletion.

A utils endpoint such as `/utils/database/reset-sequence` can be implemented for admins. It can be used when there had been direct insertions into the database. This new feature has to be implemented for all vendors, and not just Postgres, which could be in a separate PR.

File: api/src/database/helpers/sequence/dialects/postgres.ts (hunk 2 above)
Comment (jaads) [reply api/src/database/helpers/sequence/dialects/postgres.ts]: oh true.. a leftover from another approach.. 👍

File: api/src/utils/reset-sequence.ts
Comment (jaads) [reply api/src/utils/reset-sequence.ts:1]: I removed it entirely, now the helper is used in the item service directly

File: api/src/database/helpers/sequence/dialects/postgres.ts (hunk 1 above)
Comment (jaads) [reply api/src/database/helpers/sequence/dialects/postgres.ts]: I added the snippet - it now determines the sequence names dynamically 🎉

**Comment by jaads** [comment]
@joselcvarela The sequence now gets reset for each item that is created, no matter of it's origin. It's still in draft, but you could have a look anyways already. maybe you notice something.  

shall we do anything else, which is not related to the current problem, in a different PR later on? like doing a reset on deletion and an endpoint to trigger the reset manually by the user

File: tests/blackbox/routes/items/m2a.test.ts [hunk 3]
```diff
@@ -2170,6 +2170,8 @@ describe.each(common.PRIMARY_KEY_TYPES)('/items', (pkType) => {
 
 			describe('updates the auto increment value correctly', () => {
 				it.each(vendors)('%s', async (vendor) => {
+					if (vendor !== 'postgres') return;
```
Comment (licitdev) [inline tests/blackbox/routes/items/m2a.test.ts:2173]: I think we update this condition across all tests to run for all vendors except the specifically excluded ones.

Comment (jaads) [reply tests/blackbox/routes/items/m2a.test.ts]: alright!

File: api/src/services/items.ts [hunk 4]
```diff
@@ -207,6 +225,9 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 				payload[primaryKeyField] = primaryKey;
 			}
 
+			// at this point, the primary key is guaranteed to be set to the actual returned value or the best guest value
```
Comment (paescuj) [inline api/src/services/items.ts:228]: I assume
```suggestion
			// at this point, the primary key is guaranteed to be set to the actual returned value or the best guessed value
```

File: api/src/services/items.ts [hunk 5]
```diff
@@ -174,8 +174,26 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 			const payloadWithoutAliases = pick(payloadWithA2O, without(fields, ...aliases));
 			const payloadWithTypeCasting = await payloadService.processValues('create', payloadWithoutAliases);
 
-			// In case of manual string / UUID primary keys, the PK already exists in the object we're saving.
-			let primaryKey = payloadWithTypeCasting[primaryKeyField];
+			// The primary key can already exist in the object we're saving.
```
Comment (paescuj) [inline api/src/services/items.ts:177]:
```suggestion
			// The primary key can already exist in the payload.
```

File: api/src/services/items.ts [hunk 6]
```diff
@@ -174,8 +174,26 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
...
-			// In case of manual string / UUID primary keys, the PK already exists in the object we're saving.
-			let primaryKey = payloadWithTypeCasting[primaryKeyField];
+			// The primary key can already exist in the payload.
+			// In case of manual string / UUID primary keys it's always provided at this point.
+			// In case of an integer primary key, it might be provided as the user can specify the value manually.
+			let primaryKey: undefined | PrimaryKey = payloadWithTypeCasting[primaryKeyField];
+
+			// If a PK of type number was provided, although the PK is set the auto_increment,
+			// the sequence needs to be reset to protect future PK collisions.
+			// This is currently only done for PostgreSQL.
```
Comment (paescuj) [inline api/src/services/items.ts:184]:
```suggestion
```

File: api/src/services/items.ts [hunk 7]
```diff
@@ -174,8 +174,25 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
...
 
-			// In case of manual string / UUID primary keys, the PK already exists in the object we're saving.
-			let primaryKey = payloadWithTypeCasting[primaryKeyField];
+			// The primary key can already exist in the payload.
+			// In case of manual string / UUID primary keys it's always provided at this point.
+			// In case of an integer primary key, it might be provided as the user can specify the value manually.
+			let primaryKey: undefined | PrimaryKey = payloadWithTypeCasting[primaryKeyField];
+
+			// If a PK of type number was provided, although the PK is set the auto_increment,
+			// the sequence needs to be reset to protect future PK collisions.
```
Comment (paescuj) [inline api/src/services/items.ts:183]:
```suggestion
			// If a PK of type number was provided, although the PK is set the auto_increment,
			// depending on the database, the sequence needs to be reset to protect future PK collisions.
```

File: api/src/services/items.ts [hunk 8]
```diff
@@ -174,8 +174,25 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
...
 
-			// In case of manual string / UUID primary keys, the PK already exists in the object we're saving.
-			let primaryKey = payloadWithTypeCasting[primaryKeyField];
+			// The primary key can already exist in the payload.
+			// In case of manual string / UUID primary keys it's always provided at this point.
+			// In case of an integer primary key, it might be provided as the user can specify the value manually.
+			let primaryKey: undefined | PrimaryKey = payloadWithTypeCasting[primaryKeyField];
+
+			// If a PK of type number was provided, although the PK is set the auto_increment,
+			// depending on the database, the sequence needs to be reset to protect future PK collisions.
```
Comment (paescuj) [inline api/src/services/items.ts:183]:
```suggestion
			// If a PK of type number was provided, although the PK is set the auto_increment,
			// depending on the database, the sequence might need to be reset to protect future PK collisions.
```

File: api/src/services/items.ts [hunk 9]
```diff
@@ -207,6 +224,9 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 				payload[primaryKeyField] = primaryKey;
 			}
 
+			// at this point, the primary key is guaranteed to be set to the actual returned value or the best guessed value
```
Comment (paescuj) [inline api/src/services/items.ts:227]:
```suggestion
			// At this point, the primary key is guaranteed to be set.
```

**Review by paescuj — APPROVED** [review]

**Review by licitdev — APPROVED** [review]

**Pull request merged**